File: problem_5/smallest_multiple_test.rb
require 'minitest/pride'
require 'minitest/autorun'

class SmallestMultipleTest < Minitest::Test
  def smallest_multiple(max)
    range = (2..max - 1)
    num = max

    loop do
      divisible = true
      range.each { |i| break divisible = false if num % i != 0 }
      return num if divisible
      num += max
    end
  end

  def test_smallest_multiple
    assert_equal 2520, smallest_multiple(10)
    assert_equal 360360, smallest_multiple(15)
    assert_equal 232792560, smallest_multiple(20)
  end
end

File: problem_5/smallest-multiple.test.js
function smallestMultiple(maxFactor) {
  let num = maxFactor;
  while (!divisibleFactors(num, maxFactor)) {
    num += maxFactor;
  }

  return num;
}

function divisibleFactors(num, maxFactor) {
  for (let i = 1; i <= maxFactor; i++) {
    if (num % i !== 0) {
      return false;
    }
  }
  return true;
}

//----------------- Tests -----------------//

test("smallestMultiple", () => {
  expect(smallestMultiple(10)).toBe(2520);
  expect(smallestMultiple(15)).toBe(360360);
  expect(smallestMultiple(20)).toBe(232792560);
});

test("divisibleFactors", () => {
  expect(divisibleFactors(15, 3)).toBe(false);
  expect(divisibleFactors(2520, 10)).toBe(true);
  expect(divisibleFactors(360360, 15)).toBe(true);
});
